### ISSUE TITLE
Separate action for bundle install and cache

### ISSUE THREAD

**Issue by natematykiewicz** [ruby/setup-ruby#155]
In a repo at my work, we needed to install a private gem. I had to configure the auth token to get it to install. It took me a while to figure out the correct ENV to set to add our GitHub Packages credentials.

Per the GitHub Packages documentation, you're supposed to run:

```
bundle config https://rubygems.pkg.github.com/OWNER USERNAME:TOKEN
```

https://docs.github.com/en/packages/guides/configuring-rubygems-for-use-with-github-packages#authenticating-to-github-packages

The only solution I found was to set BUNDLE_RUBYGEMS__PKG__GITHUB__COM, but that doesn't set my token for just this `OWNER`, it sets the token for all packages hosted on GitHub (a bit looser than I'd like, but the only other option was the manual actions cache).

The trouble is, setup-ruby doesn't (from what I could tell) give you a way to run `bundle config` commands in setup-ruby when using the bundler cache option. I can't run bundle config before the action, because I don't have Ruby installed yet, much less bundler. But I can't run it after because it's already failed to install our private gem by then.

I feel like it would be nice if there was an action that installed the expected Ruby and Bundler versions. And a separate action that bundle installs and caches the output. That would give me the option to `bundle config` between the two.

That seems to be what you're suggesting as well. You could even keep the bundler-cache option as a shortcut (and backwards compatibility).

---

This is a continuation of the conversation started here: https://github.com/thoughtbot/shoulda-matchers/pull/1397#issuecomment-782710793

The issue I'm referring to is not related to Shoulda Matchers, so this may help 2 problems.

**Comment by eregon** [comment]
Yes, using the `BUNDLE_*` env vars is the [recommended approach](https://github.com/ruby/setup-ruby#bundle-config) now, and it's faster than running `bundle config`.
One can easily find out which variable it is with:
```
bundle config --local KEY VALUE
cat .bundle/config
```

> but that doesn't set my token for just this OWNER, it sets the token for all packages hosted on GitHub (a bit looser than I'd like, but the only other option was the manual actions cache).

I don't understand this part.
If you like you could unset/set-to-empty that variable after `bundle install`.

---

While 2 actions sounds nice conceptually I'm worried in practice it's actually quite messy because for example the "setup Ruby" part needs to include installing Bundler (so that you can run `bundle config` in between if you like), and then the "bundle install" part needs to somehow find out which version of Bundler to use, as from the `bundler` input of the "setup Ruby" part.

I've thought also about a new `bundle-config` or so input to this action, where one could set it to a few bundle config commands, and have those run in between, but it sounds like it could quickly get out of hands if people expect specific shell semantics or so for it.

**Comment by natematykiewicz** [comment]
The problem for me was the slashes:

```
$ bundle config --local https://rubygems.pkg.github.com/foo abc123
$ cat .bundle/config

BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/foo: "abc123"
```

I couldn't figure out how to set `BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/foo` as an environment variable. Best I could come up with was setting `export BUNDLE_RUBYGEMS__PKG__GITHUB__COM="abc123"`.

If I needed to run:

```
bundle config --local https://rubygems.pkg.github.com/foo abc123
bundle config --local https://rubygems.pkg.github.com/bar def456
```

I don't know how I'd set both of those via ENV, because I can't include the repo owner part of the URL, just the domain. Maybe I just need to learn how to escape the slashes?

**Comment by eregon** [comment]
Wouldn't this work?
```yaml
jobs:
  test:
    env:
      BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/foo: "abc123"
```

It seems to be parsed correctly at least by Ruby's YAML:
```
$ ruby -ryaml -e 'pp YAML.load_file ARGV[0]' test.yml 
{"jobs"=>
  {"test"=>
    {"env"=>{"BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/foo"=>"abc123"}}}}
```

One can also add quotes if wanted:
```yml
jobs:
  test:
    env:
      "BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/foo": "abc123"
```

With both Bundler 1 and 2 I see a slightly different key though:
```
$ bundle config --local https://rubygems.pkg.github.com/foo abc123
$ cat .bundle/config
---
BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/FOO/: "abc123"
```
and that works fine, as can be verified with:
```
$ env BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/FOO/="abc123" bundle config https://rubygems.pkg.github.com/foo
Settings for `https://rubygems.pkg.github.com/foo` in order of priority. The top value will be used
Set via BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/FOO/: "abc123"
```

**Comment by natematykiewicz** [comment]
Okay, after a lot of trial and error, I got it to work. It needs quotes and a trailing slash. All other attempts fail to install the gem.

```yml
jobs:
  test:
    env:
      "BUNDLE_HTTPS://RUBYGEMS__PKG__GITHUB__COM/foo/": "abc123"
```

**Comment by natematykiewicz** [comment]
At this point, I feel like maybe these findings could go into the readme? I'm fine with using the ENV var. It's simpler. I just couldn't figure out how to do it before, and wasn't 100% sure this upper snake case string was actually an ENV var or something else.

Funny enough, I see this section was added to the Readme on Jan 16th. https://github.com/ruby/setup-ruby/commit/c782da25bddec07305679a6df2495a2e77d49004#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

I switched to `bundle-config: true` on this app on January 12th.
